### PR TITLE
Fix KeyError in find_contours

### LINKED ISSUE #817: measure.find_contours Key Error Stacktrace/fail
measure.find_contours failing with a KeyError message.

I can provide the image file (png) if needed. Note the image is inverted prior to this call.

image = Image.open(imgpath)
image_inv = PIL.ImageOps.invert(image)
npai = image_inv
npaif = np.array(npai, dtype='float')
npaif_contours = measure.find_contours(npaif,level=128.0,fully_connected='low',positive_orientation='low')

Traceback (most recent call last):
  File "lib\skimagelearn.py", line 44, in <module>
    npaif_contours = measure.find_contours(npaif,level=128.0,fully_connected='low',positive_orientation='low')
  File "E:\p\pythonlib64\lib\skimage\measure\find_contours.py", line 120, in find_contours
    contours = _assemble_contours(_take_2(point_list))
  File "E:\p\pythonlib64\lib\skimage\measure\find_contours.py", line 170, in _assemble_contours
    del ends[tail[-1]]
KeyError: (0.0, 396)

### LINKED ISSUE #4497: Key error when using skimage.measure.find_contours
## Description
I am using `skimage.measure.find_contours` to extract a large number of contours from a time-series of satellite imagery represented as `numpy.arrays`. Although this works perfectly for almost all my arrays, I am receiving a `KeyError` when `skimage.measure.find_contours` is applied to certain arrays, even though I can't see anything obviously different about these layers.

My original datasets are approximately 2000 x 2000 pixels in size, with values in `float32` ranging between -1.0 and 1.0 and containing `NaN`. I have isolated the problem area down to a small 100 x 100 pixel section of the problematic array (attached as a [zip file below](https://github.com/scikit-image/scikit-image/files/4295788/array.zip)):

image

## Way to reproduce

Download and unzip 100 x 100 numpy array (`array.npy`): [array.zip](https://github.com/scikit-image/scikit-image/files/4295788/array.zip)

```python
import numpy as np
from skimage.measure import find_contours

# Load numpy array 
array = np.load('array.npy')
find_contours(array, 0)
```
Error:
```
---------------------------------------------------------------------------
KeyError                                  Traceback (most recent call last)
<ipython-input-1-de11203f72d2> in <module>
      3 
      4 array = np.load('array.npy')
----> 5 find_contours(array, 0)

/g/data/v10/public/modules/dea-env/20191127/lib/python3.6/site-packages/skimage/measure/_find_contours.py in find_contours(array, level, fully_connected, positive_orientation, mask)
    133                                                      fully_connected == 'high',
    134                                                      mask=mask)
--> 135     contours = _assemble_contours(_take_2(point_list))
    136     if positive_orientation == 'high':
    137         contours = [c[::-1] for c in contours]

/g/data/v10/public/modules/dea-env/20191127/lib/python3.6/site-packages/skimage/measure/_find_contours.py in _assemble_contours(points_iterator)
    199                     tail.extendleft(reversed(head))
    200                     # remove all traces of head:
--> 201                     del starts[head[0]]
    202                     del ends[from_point]
    203                     del contours[head_num]

KeyError: (9, 6.0)
```

## Version information

```python
3.6.7 | packaged by conda-forge | (default, Nov  6 2019, 16:19:42) 
[GCC 7.3.0]
Linux-2.6.32-754.18.2.el6.x86_64-x86_64-with-centos-6.10-Final
scikit-image version: 0.16.2
numpy version: 1.17.3

```

### PR DESCRIPTION
**Pull request by rfezzani**
## Description

Fixes #4497, Fixes #817.

In this PR, the dictionary entries deletion using `del` in `find_contours` are replaced by `dict.pop` to avoid `KeyError` raising when the key doesn't exist.

### CODE REVIEW

**Comment by jni** [comment]
@rfezzani wow, thanks! :tada: Two Qs:

(1) Could you add some tests based on the two issues this fixes?
(2) Does this fix the issue or just mask a deeper issue? e.g. I'm worried that this means we're now open to improperly-closed contours, for example.

**Comment by rfezzani** [comment]
@jni, 
(1) should I use the two images provided in the issues for the tests?
(2) the fix proposed in this PR is probably masking a non identified corner case, we may be should investigate a little bit more but the `KeyError` raising is no more an issue...

**Comment by rfezzani** [comment]
Link here to @JDWarner https://github.com/scikit-image/scikit-image/issues/817#issuecomment-76540861.

**Comment by rfezzani** [comment]
And to @zpincus https://github.com/scikit-image/scikit-image/issues/817#issuecomment-76892791 that somehow predicted the issue #4497 :clap:

**Comment by rfezzani** [comment]
Here is the output of the example of @robbibt
contours
We in fact observe two strange contours (red and green)...

**Comment by jni** [comment]
@rfezzani great finds! (those comments) there's a reason those two sit on our steering council! :joy: 

And also good job with the test image. I guess we need to figure out why those contours end up like that. Seeing the contours makes me think we could isolate the problem to a tiny segment of the image and follow every part of the loop.

**Comment by rfezzani** [comment]
Unfortunately, the strange contours seems not to be directly connected to the `KeyError` raise. In fact, `find_contours` ends correctly on the top left area containing the strange contours:
```python
>>> import skimage
>>> skimage.__version__
'0.16.2'
>>> array = np.load('array.npy')
>>> from skimage.measure import find_contours
>>> c = find_contours(array[:30, :30], 0)
>>> # Supposed to crash here!!!>>> c = find_contours(array, 0)
>>> c = find_contours(array, 0)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "~~~~~~~~~~~~~~~/lib/python3.7/site-packages/skimage/measure/_find_contours.py", line 135, in find_contours
    contours = _assemble_contours(_take_2(point_list))
  File "~~~~~~~~~~~~~~~/lib/python3.7/site-packages/skimage/measure/_find_contours.py", line 201, in _assemble_contours
    del starts[head[0]]
KeyError: (9, 6.0)
```
and the strange contours are still there:
contours_small

**Comment by rfezzani** [comment]
OK, the strange contours are valid contours... Here is the result when we replace the nans contained in the array by zeros (contour's level):
contours_small_no_nan
They only look strange because they are on the border of the masked array...

**Comment by rfezzani** [comment]
OK, I finally got it: The problem was that the contour's `starts` and `ends` dictionary was not up to date!
I added a full update of these two dictionaries when two contours are joined, and that solved the problem! To highlight this, I allowed the `KeyError` to be raised again, and the examples from #4497 and # 817 pass without any error!

File: skimage/measure/_find_contours.py [hunk 1]
```diff
@@ -114,47 +114,38 @@ def find_contours(array, level,
...
-       positive_orientation not in _param_options):
-        raise ValueError('Parameters "fully_connected" and'
-        ' "positive_orientation" must be either "high" or "low".')
-    point_list = _find_contours_cy.iterate_and_store(array, level,
-                                                     fully_connected == 'high',
-                                                     mask=mask)
-    contours = _assemble_contours(_take_2(point_list))
+        if not np.can_cast(mask.dtype, bool, casting='safe'):
+            raise TypeError('Parameter "mask" must be a binary array.')
+        mask = mask.astype(np.uint8)
```
Comment (jni) [inline skimage/measure/_find_contours.py]:
```suggestion
        mask = mask.astype(np.uint8, copy=False)
```

**Comment by rfezzani** [comment]
I performed a refactoring of the `find_contour` function:

- all input validations are moved at the begin of the function to exit as soon as possible in case of problem.
- in `_assemble_contours`, the [`_dict.get(key)`](https://github.com/scikit-image/scikit-image/blob/master/skimage/measure/_find_contours.py#L166) followed by [`del _dict[key]`](https://github.com/scikit-image/scikit-image/blob/master/skimage/measure/_find_contours.py#L178) are replaced by a [`_dict.pop(key)`](https://github.com/scikit-image/scikit-image/pull/4505/files#diff-fdc7c9210712d9f5366d8a0cdeb92863R157). This allows to get rid of the definition of the intermediary variables `tail_data` and `head_data`.
- [`iterate_and_store`](https://github.com/scikit-image/scikit-image/blob/master/skimage/measure/_find_contours_cy.pyx#L18) now returns segments (couples of points) instead of a list of points. The [`_take_2`](https://github.com/scikit-image/scikit-image/blob/master/skimage/measure/_find_contours.py#L141) function is no more needed.
- The [loop]( https://github.com/scikit-image/scikit-image/blob/master/skimage/measure/_find_contours_cy.pyx#L60) over image pixels is simplified (use nested loop instead of defining [`coord`](https://github.com/scikit-image/scikit-image/blob/master/skimage/measure/_find_contours_cy.pyx#L47) argument and [checking](https://github.com/scikit-image/scikit-image/blob/master/skimage/measure/_find_contours_cy.pyx#L96) its values).
- `iterate_and_store` is renamed to `_get_contour_segments`.

**Comment by zpincus** [comment]
Yes, thank you, @rfezzani! This is a really nice cleanup of code that I originally wrote years ago as a grad student and which I've been sadly neglectful of since. This refactor is much more elegant for sure :)

**Comment by emmanuelle** [comment]
Thank you @rfezzani this is awesome :-). Merging!

**Comment by robbibt** [comment]
Hi all, has this update been rolled into version `0.17.2` of `skimage`? I just encountered the same `Key error` issue on `0.17.2` that I originally raised in #4497, and I was wondering if the fix hasn't been released yet, or if this is something new (I can put together a reproducible example if it's the second option):

```
---------------------------------------------------------------------------
KeyError                                  Traceback (most recent call last)
/g/data1a/r78/DEACoastLines/deacoastlines_statistics.py in contours_to_multiline(da_i, z_value, min_vertices)
    136             line_features = [LineString(i[:,[1, 0]]) 
--> 137                              for i in find_contours(da_i.data, z_value)
    138                              if i.shape[0] > min_vertices]

/g/data/v10/public/modules/dea-env/20200612/lib/python3.6/site-packages/skimage/measure/_find_contours.py in find_contours(array, level, fully_connected, positive_orientation, mask)
    136                                      fully_connected == 'high', mask=mask)
--> 137     contours = _assemble_contours(segments)
    138     if positive_orientation == 'high':

/g/data/v10/public/modules/dea-env/20200612/lib/python3.6/site-packages/skimage/measure/_find_contours.py in _assemble_contours(segments)
    173                     # remove all traces of tail:
--> 174                     ends.pop(tail[-1])
    175                     contours.pop(tail_num, None)

KeyError: (1452.0, 1086)

During handling of the above exception, another exception occurred:

KeyError                                  Traceback (most recent call last)
<ipython-input-7-96fe0e1e6fe8> in <module>
      4                                              min_vertices=30,
      5                                              dim='year',
----> 6                                              output_path='temp.geojson')
      7                 .set_index('year'))

/g/data1a/r78/DEACoastLines/deacoastlines_statistics.py in subpixel_contours(da, z_values, crs, affine, attribute_df, output_path, min_vertices, dim, errors)
    194         contour_arrays = {str(i)[0:10]: 
    195                           contours_to_multiline(da_i, z_values[0], min_vertices)
--> 196                           for i, da_i in da.groupby(dim)}
    197 
    198     # If attributes are provided, add the contour keys to that dataframe

/g/data1a/r78/DEACoastLines/deacoastlines_statistics.py in <dictcomp>(.0)
    194         contour_arrays = {str(i)[0:10]: 
    195                           contours_to_multiline(da_i, z_values[0], min_vertices)
--> 196                           for i, da_i in da.groupby(dim)}
    197 
    198     # If attributes are provided, add the contour keys to that dataframe

/g/data1a/r78/DEACoastLines/deacoastlines_statistics.py in contours_to_multiline(da_i, z_value, min_vertices)
    141             line_features = [LineString(i[:,[1, 0]]) 
    142                              for i in find_contours(da_i.data, z_value, 
--> 143                                                     fully_connected='high') 
    144                              if i.shape[0] > min_vertices]          
    145 

/g/data/v10/public/modules/dea-env/20200612/lib/python3.6/site-packages/skimage/measure/_find_contours.py in find_contours(array, level, fully_connected, positive_orientation, mask)
    135     segments = _get_contour_segments(array.astype(np.double), float(level),
    136                                      fully_connected == 'high', mask=mask)
--> 137     contours = _assemble_contours(segments)
    138     if positive_orientation == 'high':
    139         contours = [c[::-1] for c in contours]

/g/data/v10/public/modules/dea-env/20200612/lib/python3.6/site-packages/skimage/measure/_find_contours.py in _assemble_contours(segments)
    172                     head.extend(tail)
    173                     # remove all traces of tail:
--> 174                     ends.pop(tail[-1])
    175                     contours.pop(tail_num, None)
    176                     # Update contour starts end ends

KeyError: (1452.0, 1086)
```

**Comment by robbibt** [comment]
I'm fairly certain this is a bug - reproducible example here:
```
import numpy as np
from skimage.measure import find_contours

# Example array
array = np.array([[ 0.08, -0.03, -0.17,  -0.08,  0.24,  0.06,  0.17, -0.02],
                  [ 0.12,  0.  ,  np.nan, 0.24,  0.  , -0.53,  0.26,  0.16],
                  [ 0.39,  0.  ,  0.  ,   0.  ,  0.  , -0.02, -0.3 ,  0.01],
                  [ 0.28, -0.04, -0.03,   0.16,  0.12,  0.01, -0.87,  0.16],
                  [ 0.26,  0.08,  0.08,   0.08,  0.12,  0.13,  0.11,  0.19],
                  [ 0.27,  0.24,  0.  ,   0.25,  0.32,  0.19,  0.26,  0.22]],
      dtype=np.float32)

# Extract 0 contour
find_contours(array, 0)
```
```
---------------------------------------------------------------------------
KeyError                                  Traceback (most recent call last)
<ipython-input-138-952e103d773c> in <module>
     12 
     13 # Extract 0 contour
---> 14 find_contours(array, 0)

/g/data/v10/public/modules/dea-env/20200612/lib/python3.6/site-packages/skimage/measure/_find_contours.py in find_contours(array, level, fully_connected, positive_orientation, mask)
    135     segments = _get_contour_segments(array.astype(np.double), float(level),
    136                                      fully_connected == 'high', mask=mask)
--> 137     contours = _assemble_contours(segments)
    138     if positive_orientation == 'high':
    139         contours = [c[::-1] for c in contours]

/g/data/v10/public/modules/dea-env/20200612/lib/python3.6/site-packages/skimage/measure/_find_contours.py in _assemble_contours(segments)
    172                     head.extend(tail)
    173                     # remove all traces of tail:
--> 174                     ends.pop(tail[-1])
    175                     contours.pop(tail_num, None)
    176                     # Update contour starts end ends

KeyError: (2.0, 3)
```

Can someone re-open #4497, or should I create a new issue? @rfezzani @jni

**Comment by jni** [comment]
@robbibt **sigh**, thank you for raising this! This PR was indeed in 0.17, so this is a "new" bug. Would you mind raising a new issue with this example? (And link here.) We'll indeed have to investigate again.

Thanks and sorry that the first fix wasn't quite sufficient! :grimacing:

**Comment by robbibt** [comment]
@jni No worries, here we go! https://github.com/scikit-image/scikit-image/issues/4830
